third_party/terraform/resources/resource_storage_bucket.go

```diff
@@ -755,89 +768,162 @@ func flattenBucketLifecycleRuleCondition(condition *storage.BucketLifecycleRuleC
-			lifecycle_rule := raw_lifecycle_rule.(map[string]interface{})
+func expandStorageBucketLifecycle(v interface{}) (*storage.BucketLifecycle, error) {
+	if v == nil {
+		return &storage.BucketLifecycle{
+			ForceSendFields: []string{"Rule"},
+		}, nil
+	}
+	lifecycleRules := v.([]interface{})
+	transformedRules := make([]*storage.BucketLifecycleRule, 0, len(lifecycleRules))
 
-			target_lifecycle_rule := &storage.BucketLifecycleRule{}
+	for _, v := range lifecycleRules {
+		rule, err := expandStorageBucketLifecycleRule(v)
+		if err != nil {
+			return nil, err
+		}
+		transformedRules = append(transformedRules, rule)
+	}
 
-			if v, ok := lifecycle_rule["action"]; ok {
-				if actions := v.(*schema.Set).List(); len(actions) == 1 {
-					action := actions[0].(map[string]interface{})
+	if len(transformedRules) == 0 {
+		return &storage.BucketLifecycle{
+			ForceSendFields: []string{"Rule"},
+		}, nil
+	}
 
-					target_lifecycle_rule.Action = &storage.BucketLifecycleRuleAction{}
+	return &storage.BucketLifecycle{
+		Rule: transformedRules,
+	}, nil
+}
 
-					if v, ok := action["type"]; ok {
-						target_lifecycle_rule.Action.Type = v.(string)
-					}
+func expandStorageBucketLifecycleRule(v interface{}) (*storage.BucketLifecycleRule, error) {
+	if v == nil {
+		return nil, nil
+	}
 
-					if v, ok := action["storage_class"]; ok {
-						target_lifecycle_rule.Action.StorageClass = v.(string)
-					}
-				} else {
-					return fmt.Errorf("Exactly one action is required")
-				}
-			}
+	rule := v.(map[string]interface{})
+	transformed := &storage.BucketLifecycleRule{}
 
-			if v, ok := lifecycle_rule["condition"]; ok {
-				if conditions := v.(*schema.Set).List(); len(conditions) == 1 {
-					condition := conditions[0].(map[string]interface{})
+	if v, ok := rule["action"]; ok {
+		action, err := expandStorageBucketLifecycleRuleAction(v)
+		if err != nil {
+			return nil, err
+		}
+		transformed.Action = action
+	} else {
+		return nil, fmt.Errorf("exactly one action is required for lifecycle_rule")
+	}
 
-					target_lifecycle_rule.Condition = &storage.BucketLifecycleRuleCondition{}
+	if v, ok := rule["condition"]; ok {
+		cond, err := expandStorageBucketLifecycleRuleCondition(v)
+		if err != nil {
+			return nil, err
+		}
+		transformed.Condition = cond
+	}
 
-					if v, ok := condition["age"]; ok {
-						target_lifecycle_rule.Condition.Age = int64(v.(int))
-					}
+	return transformed, nil
+}
 
-					if v, ok := condition["created_before"]; ok {
-						target_lifecycle_rule.Condition.CreatedBefore = v.(string)
-					}
+func expandStorageBucketLifecycleRuleAction(v interface{}) (*storage.BucketLifecycleRuleAction, error) {
+	if v == nil {
+		return nil, fmt.Errorf("exactly one action is required for lifecycle_rule")
+	}
 
-					if v, ok := condition["is_live"]; ok {
-						target_lifecycle_rule.Condition.IsLive = googleapi.Bool(v.(bool))
-					}
+	actions := v.(*schema.Set).List()
+	if len(actions) != 1 {
```

Comment:
I'm confused why action is a Set with a max size of 1.

Reply:
probably historical, I think there was a point we tried representing nested objects as sets with maxitems: 1 instead of as lists.

Reply:
Condition is also the same (max # = 1). I think it's because usually when we did it, we wanted to support the possible situation of multiple actions per condition or similar? It probably should be a list for action since I doubt people are keen to have a list of actions applied in arbitrary order, but it would make sense for multiple conditions.